### ISSUE TITLE
FIRESTORE INTERNAL ASSERTION FAILED: closeWithFinalState

### ISSUE THREAD

**Issue by x0rb0t** [firebase/firebase-ios-sdk#596]
> Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'FIRESTORE INTERNAL ASSERTION FAILED: closeWithFinalState should only be called for a started stream that has an active delegate.'
> *** First throw call stack:
> (0x1848fa364 0x183b40528 0x1848fa238 0x1852957f4 0x10041b724 0x10040e7d4 0x10040e8dc 0x1003db690 0x1003d1c00 0x101d0d2cc 0x101d0d28c 0x101d1bf80 0x101d107ec 0x101d1cf6c 0x101d24020 0x18451ef1c 0x18451eb6c)
> libc++abi.dylib: terminating with uncaught exception of type NSException

It happens if:
 1) You has login: 
`Auth.auth().signIn(with: credential) { (user, error) in
                if let _ = error {
                    return
                }
                self.mainView?.didLogin()
                self.dismiss(animated: true, completion: nil)
}`
2) Perform write to firestore like this:
`Firestore.firestore().collection("users").document(currentUser.uid).updateData(data)
        { (err) in
            if let err = err {
                print("Failed to save location info into db:", err)
                return
            }
            print("Successfully saved location info to db")
        }`
3) You perform sign out "try Auth.auth().signOut()"
4) Login again. Got error.
5) If skip step 2 - no error

Device: Iphone 5s 
xCode: Version 9.2 (9C40b)
Using BoringSSL (9.1)
Using Firebase (4.8.0)
Using FirebaseAnalytics (4.0.5)
Using FirebaseAuth (4.4.1)
Using FirebaseCore (4.0.13)
Using FirebaseDatabase (4.1.3)
Using FirebaseFirestore (0.9.4)
Using FirebaseInstanceID (2.0.8)
Using FirebaseMessaging (2.0.8)
Using FirebaseStorage (2.1.1)
Using GTMSessionFetcher (1.1.12)
Using GoogleToolboxForMac (2.1.3)
Using Protobuf (3.5.0)
Using gRPC (1.8.0)
Using gRPC-Core (1.8.0)
Using gRPC-ProtoRPC (1.8.0)
Using gRPC-RxLibrary (1.8.0)
Using leveldb-library (1.20)
Using nanopb (0.3.8)
Target ios version: 10.0

**Comment by x0rb0t** [comment]
On iphone 6s Emulator error doesn't happens

**Comment by antonioallen** [comment]
I am also receiving this error.

**Comment by AchidFarooq** [comment]
Same problem here, dont know what the solution is.

**Comment by uknowmeright** [comment]
Same thing. It only happens for the google Sign in for me. If I sign in with the following after google gets the credential:

Auth.auth().signIn(with: credential) { (user, error) in

Then try signing out with:

```
        let firebaseAuth = Auth.auth()
        do {
            try firebaseAuth.signOut()
        } catch let signOutError as NSError {
            print ("Error signing out: %@", signOutError)
        }
```

It crashes with the following error:

NSInternalInconsistencyException', reason: 'FIRESTORE INTERNAL ASSERTION FAILED: closeWithFinalState should only be called for a started stream that has an active delegate

**Comment by AchidFarooq** [comment]
Could it be possible that we have to add `// Stop listening to changes` ? Maybe we are making too much networkrequests?

**Comment by yaronalk** [comment]
Same here.
Does anybody know what's the last "safe" version of the SDK that I can revert to?

**Comment by AchidFarooq** [comment]
It's really weird, but i haven't had 1 crash today. Seems to be fixed?

**Comment by morganchen12** [comment]
@mikelehen, this looks like something that may be fixed by the fix in your [comment](https://github.com/firebase/firebase-ios-sdk/issues/633#issuecomment-356444265) in #633.

**Comment by mikelehen** [comment]
I need to sync with @wilhuff and double-check on the status here. I thought this was diagnosed and resolved, but I may have been mistaken...  hold tight.

**Comment by mikelehen** [comment]
Ok. We believed this was fixed by https://github.com/firebase/firebase-ios-sdk/pull/510 but upon closer inspection, that was included in 0.9.4 and so should no longer be getting hit. So I'll dig into this.

**Comment by mikelehen** [comment]
Does anybody have a reliable way to reproduce this? I haven't been able to trigger it. I tried a straightforward sign in, do a write, sign out, sign in again as described in the issue, but that worked fine. So then I tried doing writes every 50ms and then signing in then out over and over, and that worked fine (although auth gets angry if you sign in too many times too quickly).

So I'm not able to reproduce this yet. If anybody has a full code snippet that reproduces the crash or any other hints, please let me know.

**Comment by Ruberik** [comment]
Mostly useless hint: this happens to me sometimes if I leave the simulator running on my laptop. About half the time, if I come back to it after at least a few hours, I get this error. That's on the iPhone 6 emulator, Firebase 4.8.0, Firestore 0.9.4. I'm guessing the laptop gets put to sleep and woken up, but there wouldn't be any human interaction with the emulator in that time.

**Comment by mikelehen** [comment]
That was actually very helpful. :-) I think I have a repro now.

**Comment by morganchen12** [comment]
Reopening until the fix is released.

**Comment by paulb777** [comment]
Fixed in Firebase 4.8.1 and FirebaseFirestore 0.10.0
